File: location_based_game/frontend/lib/models/question.dart
/// Represents a single multiple-choice question belonging to a [Topic].
class Question {
  /// Difficulty level of the question, from 1 (easy) to 5 (hard).
  final int difficulty;

  /// The question prompt.
  final String text;

  /// The list of possible answers. Must contain between 2 and 5 options.
  final List<String> options;

  /// Index of the correct answer within [options].
  final int correctOptionIndex;

  /// Explanation of the correct answer.
  final String explanation;

  Question({
    required this.difficulty,
    required this.text,
    required this.options,
    required this.correctOptionIndex,
    required this.explanation,
  })  : assert(
          difficulty >= 1 && difficulty <= 5,
          'difficulty must be between 1 and 5 (was $difficulty)',
        ),
        assert(
          options.length >= 2 && options.length <= 5,
          'options must contain between 2 and 5 elements (was ${options.length})',
        ),
        assert(
          correctOptionIndex >= 0 && correctOptionIndex < options.length,
          'correctOptionIndex must be a valid index of options',
        );

  /// Convenience getter returning the text of the correct answer.
  String get correctAnswer => options[correctOptionIndex];

  /// Returns true if [optionIndex] corresponds to the correct answer.
  bool isCorrect(int optionIndex) => optionIndex == correctOptionIndex;

  factory Question.fromJson(Map<String, dynamic> json) {
    return Question(
      difficulty: json['difficulty'] as int,
      text: json['text'] as String,
      options: List<String>.from(json['options'] as List),
      correctOptionIndex: json['correctOptionIndex'] as int,
      explanation: json['explanation'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'difficulty': difficulty,
      'text': text,
      'options': options,
      'correctOptionIndex': correctOptionIndex,
      'explanation': explanation,
    };
  }

  @override
  String toString() => 'Question(difficulty: $difficulty, text: $text)';
}

File: location_based_game/backend/src/main/java/it/unicam/locationbasedgame/model/Question.java
package it.unicam.locationbasedgame.model;

import jakarta.persistence.CollectionTable;
import jakarta.persistence.Column;
import jakarta.persistence.ElementCollection;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.OrderColumn;
import jakarta.persistence.Table;
import jakarta.persistence.Transient;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.util.List;

/**
 * Represents a single multiple-choice question belonging to a Topic.
 */
@Entity
@Table(name = "questions")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class Question {

    /** Primary key. */
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /** Difficulty level of the question. */
    @Column(nullable = false)
    private int difficulty;

    /** The question prompt. */
    @Column(nullable = false, length = 1000)
    private String text;

    /**
     * The list of possible answers. Must contain between 2 and 5 options.
     */
    @ElementCollection
    @CollectionTable(name = "question_options", joinColumns = @JoinColumn(name = "question_id"))
    @OrderColumn(name = "option_index")
    @Column(name = "option_text", nullable = false)
    private List<String> options;

    /** Index of the correct answer within options. */
    @Column(nullable = false)
    private int correctOptionIndex;

    /** Explanation of the correct answer. */
    @Column(nullable = false, length = 2000)
    private String explanation;

    /**
     * Gets the text of the correct answer.
     *
     * @return the option text located at correctOptionIndex
     */
    @Transient
    public String getCorrectAnswer() {
        return options.get(correctOptionIndex);
    }

    /**
     * Checks whether the given option index corresponds to the correct
     * answer.
     *
     * @param optionIndex the index to check
     * @return true if optionIndex is the correct answer, false otherwise
     */
    public boolean isCorrect(int optionIndex) {
        return optionIndex == correctOptionIndex;
    }
}
